crates/core/src/table/builder.rs

```diff
@@ -0,0 +1,288 @@
+        let hudi_options = self.hudi_options.unwrap_or_default().clone();
+        let mut storage_options = self.storage_options.unwrap_or_default().clone();
+
+        Self::load_storage_options(&mut storage_options);
+
+        let hudi_configs =
+            Self::load_hudi_configs(self.base_uri.clone(), hudi_options, &storage_options)
+                .await
+                .context("Failed to load table properties")?;
+
+        let hudi_configs = Arc::from(hudi_configs);
+        let storage_options = Arc::from(storage_options);
+
+        let timeline = Timeline::new(hudi_configs.clone(), storage_options.clone())
+            .await
+            .context("Failed to load timeline")?;
+
+        let file_system_view = FileSystemView::new(hudi_configs.clone(), storage_options.clone())
+            .await
+            .context("Failed to load file system view")?;
+
+        Ok(Table {
+            hudi_configs,
+            storage_options,
+            timeline,
+            file_system_view,
+        })
+    }
+
+    fn load_storage_options(storage_options: &mut HashMap<String, String>) {
+        Self::imbue_cloud_env_vars(storage_options);
+    }
+
+    async fn load_hudi_configs(
+        base_uri: String,
+        mut hudi_options: HashMap<String, String>,
+        storage_options: &HashMap<String, String>,
+    ) -> anyhow::Result<HudiConfigs> {
+        hudi_options.insert(
+            HudiTableConfig::BasePath.as_ref().to_string(),
+            base_uri.to_string(),
+        );
+
+        // create a [Storage] instance to load properties from storage layer.
+        let storage = Storage::new(
+            Arc::new(storage_options.clone()),
+            Arc::new(HudiConfigs::new(&hudi_options)),
+        )?;
+
+        Self::imbue_table_properties(&mut hudi_options, storage.clone()).await?;
+
+        Self::imbue_global_hudi_configs_if_not_present(&mut hudi_options, storage.clone()).await?;
+
+        let hudi_configs = HudiConfigs::new(hudi_options);
+
+        Table::validate_configs(&hudi_configs).expect("Hudi configs are not valid.");
+        Ok(hudi_configs)
+    }
+
+    fn imbue_cloud_env_vars(options: &mut HashMap<String, String>) {
+        const PREFIXES: [&str; 3] = ["AWS_", "AZURE_", "GOOGLE_"];
+
+        for (key, value) in env::vars() {
+            if PREFIXES.iter().any(|prefix| key.starts_with(prefix))
+                && !options.contains_key(&key.to_ascii_lowercase())
+            {
+                options.insert(key.to_ascii_lowercase(), value);
+            }
+        }
+    }
+
+    async fn imbue_table_properties(
+        options: &mut HashMap<String, String>,
+        storage: Arc<Storage>,
+    ) -> anyhow::Result<()> {
+        let bytes = storage.get_file_data(".hoodie/hoodie.properties").await?;
+        let table_properties = parse_data_for_options(&bytes, "=")?;
+
+        // TODO: handle the case where the same key is present in both table properties and options
```

Comment:
in such case, which option should take precedence?
@xushiyan

Reply:
it's a bit complicated, there are difference cases pertaining to individual config - i'll fix this todo